sql: fix auto-retry behavior with autocommit_before_ddl by buffering notices

rafiss opened the pull request:
When the autocommit_before_ddl setting was enabled, we were sending a notice to the client without any buffering. This prevents auto-retry logic for retriable errors from kicking in, since if results have already sent to the client, it's not safe to retry the current statement.

The fix is to buffer the notice for sending instead. It will get sent whenever results are flushed back to the client. In order to achieve this, I changed the SendClientNotice function so that it does not immediately flush the notice. Instead, the notice will now get sent whenever the connection is flushed, like when the results are complete or when a Flush message is received.

The existing BufferClientNotice function is not sufficient since that only buffers notices in the command result, and that buffer is discarded due to how the connExecutor state transitions are defined. Namely, the connExecutor will execute the schema change command twice: once to autocommit the current transaction, and another time to run the schema change. To avoid losing the notice from the autocommit, it must be buffered all the way into the client connection, and not just the command result.

Running with this patch makes logic tests less flaky, since schema changes that encounter retryable errors are way more likely to be able to be automatically retried now. This allows us to remove the testing knob that overrode the transaction liveness threshold. I verified the flakiness is gone by using:
```
./dev testlogic ccl --config=3node-tenant --stress --ignore-cache
```

fixes https://github.com/cockroachdb/cockroach/issues/133180
Release note (bug fix): Fixed a bug that prevented transaction retry errors encountered during implicit transactions from being automatically retried internally if the autocommit_before_ddl session variable was enabled and the statement was a schema change.

rafiss committed:
sql: fix auto-retry behavior with autocommit_before_ddl

When the autocommit_before_ddl setting was enabled, we were
sending a notice to the client without any buffering. This prevents
auto-retry logic for retriable errors from kicking in, since if results
have already sent to the client, it's not safe to retry the current
statement.

The fix is to buffer the notice for sending instead. It will get sent
whenever results are flushed back to the client. In order to achieve
this, I changed the SendClientNotice function so that it does not
immediately flush the notice. Instead, the notice will now get sent
whenever the connection is flushed, like when the results are complete
or when a Flush message is received.

The existing BufferClientNotice function is not sufficient since that only
buffers notices in the command result, and that buffer is discarded due
to how the connExecutor state transitions are defined. Namely, the
connExecutor will execute the schema change command twice: once to
autocommit the current transaction, and another time to run the schema
change. To avoid losing the notice from the autocommit, it must be
buffered all the way into the client connection, and not just the
command result.

Running with this patch makes logic tests less flaky, since schema
changes that encounter retryable errors are way more likely to be able
to be automatically retried now. This allows us to remove the testing
knob that overrode the transaction liveness threshold. I verified the
flakiness is gone by using:
```
./dev testlogic ccl --config=3node-tenant --stress --ignore-cache
```

Release note (bug fix): Fixed a bug that prevented transaction retry
errors encountered during implicit transactions from being automatically
retried internally if the autocommit_before_ddl session variable was
enabled and the statement was a schema change.

fqazi reviewed (approved):
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> 

> Reviewed 7 of 7 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/141369#-:-OIvHLUU3CMptEewwfnt:b-uwkowd)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @rafiss)

___
*[`pkg/sql/notice.go` line 34 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/141369#-OIvN_ROEC4REoLAAyBN:-OIvN_ROEC4REoLAAyBO:b-a57cf0) ([raw file](https://github.com/cockroachdb/cockroach/blob/a72503660a2bff517fcbcc4d58a623c3a53c62d9/pkg/sql/notice.go#L34)):*
> ```Go
> 	BufferNotice(pgnotice.Notice)
> 	// SendNotice sends the given notice to the client. The notice will be in
> 	// the client communication buffer until it is flushed.
> ```

Does this mean it gets sent when ever any replies are sent to client?


<!-- Sent from Reviewable.io -->

rafiss committed:
sql: use BufferNotice instead of SendNotice when possible

These usages had no need to immediately send a notice to the client.

Release note: None

rafiss reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/141369)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @fqazi)

___
*[`pkg/sql/notice.go` line 34 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/141369#-OIvN_ROEC4REoLAAyBN:-OIvyLiZ7vyH_cdOQsSD:b-2qumj6) ([raw file](https://github.com/cockroachdb/cockroach/blob/a72503660a2bff517fcbcc4d58a623c3a53c62d9/pkg/sql/notice.go#L34)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

Does this mean it gets sent when ever any replies are sent to client?
</blockquote></details>

yeah, exactly. or if the client sends a pgwire `Sync` message


<!-- Sent from Reviewable.io -->

rafiss reviewed (commented):
tftr!

bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/141369#-:-OIvyQdq2LHV9JR4jPDx:bjpalj3)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @fqazi)



<!-- Sent from Reviewable.io -->
